Devscord.DiscordFramework/Framework/Commands/Services/BotCommandsMatchingService.cs

```diff
@@ -1,87 +1,147 @@
+                .ToList();
+
+            this.CheckQuotationMarksInListsAndTexts(template.Properties, argsAndValues);
+            return this.ComparePropertiesToArgsAndValues(template.Properties, argsAndValues);
+        }
+
+        public bool AreCustomCommandArgumentsCorrect(BotCommandTemplate template, Regex customTemplate, string input)
+        {
+            var matchGroups = customTemplate.Match(input).Groups;
+            var requiredProperties = template.Properties.Where(x => !x.IsOptional);
+            if (requiredProperties.Any(x => !matchGroups.ContainsKey(x.Name)))
             {
-                return false;
+                Log.Warning("Custom template {customTemplate} is not valid for {commandName}", customTemplate, template.CommandName);
+                throw new InvalidArgumentsException();
             }
-            return true;
+            var argsAndValues = matchGroups.Keys
+                .Where(arg => !arg.All(char.IsDigit))
+                .Select(arg => new KeyValuePair<string, string>(arg, matchGroups[arg].Value.Trim()))
+                .ToList();
+
+            this.CheckQuotationMarksInListsAndTexts(template.Properties, argsAndValues, isCommandMatchedWithCustom: true);
+            return this.ComparePropertiesToArgsAndValues(template.Properties, argsAndValues);
         }
 
-        private bool CompareArgumentsToProperties(IReadOnlyCollection<DiscordRequestArgument> arguments, IReadOnlyCollection<BotCommandProperty> properties)
+        private bool AreAllGivenArgsForCommandKnown(IEnumerable<BotCommandProperty> properties, IEnumerable<DiscordRequestArgument> arguments)
         {
-            var notOptionalCount = properties.Count(x => !x.IsOptional);
-            var parametersCount = arguments.Count(x => !string.IsNullOrEmpty(x.Name));
-            if (parametersCount > properties.Count || parametersCount < notOptionalCount)
+            var argumentsNames = arguments
+                .Select(x => x.Name?.ToLowerInvariant())
+                .ToList();
+            var lists = properties
+                .Where(property => property.Type == BotCommandPropertyType.List)
+                .Select(x => x.Name.ToLowerInvariant())
+                .ToList();
+            var argsWithoutListSubarguments = this.GetArgsWithoutListSubarguments(argumentsNames, lists);
+            return argsWithoutListSubarguments.All(arg => properties.Any(property => arg == property.Name.ToLowerInvariant()));
+        }
+
+        private List<string> GetArgsWithoutListSubarguments(List<string> argumentsNames, List<string> listsNames)
+        {
+            for (int i = 0; i < argumentsNames.Count; i++)
             {
-                return false;
+                if (!listsNames.Contains(argumentsNames[i]))
+                {
+                    continue;
+                }
+                while (i + 1 < argumentsNames.Count && argumentsNames[i + 1] == null)
```

Comment:
dlaczego tu jest while a nie if?

Reply:
gdyby użytkownik wpisałby listę z jednym lub dwoma argumentami to mogłoby być if, ale może być tak, że użytkownik poda listę z trzema, czterema argumentami lub poda jeszcze większą ilość, wtedy trzeba kodu, który te wszystkie argumenty usunie. Tutaj akurat użyłem kodu, który iteracyjnie usuwa te podargumenty listy. Jako przykład, gdzie to się sprawdzi można podać taką komendę "-addrole -roles abc qwerty 123 456 789 000 asd", wtedy while będzie konieczny.